accepted/wasmtime-rpc.md

```diff
@@ -0,0 +1,162 @@
+- Integrations with some services may not be possible to be implemented in `wasmtime` or even as part of a Bytecode Alliance project due to licensing incompatibilities.
+
+From perspective of `wasmtime` embedders, host interface implementations are tightly coupled with `wasmtime` version and so are a significant maintenance burden, especially if maintained outside of `wasmtime` tree.
+`wasmtime` linker API as well as `wasmtime` WASI implementation are effectively the API surface that developers must target to extend `wasmtime` embeddings.
+Maintainers of, for example, Rust crates that provide host interface implementations require a release cadence synchronised with `wasmtime` with each release being a breaking change due to `wasmtime` crate version increasing the major version.
+
+Even though WebAssembly components become more and more capable over time, the need for direct access to host functionality (e.g. access to hardware devices) is unlikely to become redundant any time soon.
+
+Plugin use case requirements vary greatly - for example, while for some users running plugins as part of their WebAssembly runtime process may be acceptable, for others it may not be.
+
+See https://github.com/bytecodealliance/wasmtime/issues/7348 for additional context on the CLI use case
+
+# Proposal
+[proposal]: #proposal
+
+The proposal is to allow extending `wasmtime` CLI with out-of-tree, RPC-based host plugins and use [wRPC] as the (first, potentially out of many) RPC protocol.
+
+This would be an *addition* to existing functionality, rather than a replacement of it, which would compose with what is already available.
+
+The CLI `run` and `serve` commands would support an additional option, it seems that `-P` for `--plugin` would fit well:
+
+```
+  -P, --plugin <KEY[=VAL[,..]]>
+          Plugin-related configuration options, `-P help` to see all
+```
+
+```
+Available plugin options:
+
+  -P protocol=<protocol> --
+
+        Selects the plugin protocol to use, e.g. `wrpc+tcp` or `wrpc+unix`
+  -P target=<target> --
+
+        Specifies the protocol-specific plugin target, which provides
+        imports.
+        For `wrpc+tcp`, this is a resolvable address, e.g. `localhost:7761`
+        For `wrpc+unix`, this is a path on the file system e.g. `/tmp/wasmtime-rpc`
+  -P import=<spec>[,spec[,..]] --
+
+        Selects the interfaces to import using the plugin.
+        The value is a sequence comma-separated interface specifiers.
+        For example: `wasi:keyvalue` or `wasi:keyvalue,wasi:nn`
+```
+
+Example usage:
+
+```
+wasmtime run -P protocol=wrpc+tcp -P target=[::1]:7761 -P import=wasi:keyvalue,wasi:nn my.wasm
+```
+
+To improve usability, Wasmtime could standardize on default socket address or Unix domain socket path to use.
+For example, I chose port `7761` as the default for TCP, which is specified as "unassigned" in [IANA registry](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?&page=108).
```

Comment:
If a default port is used, should it be registered with IANA? And what should the default be for unix domain sockets? `/run/user/<uid>/<package_name>.sock`?

Reply:
I chose `7761` pretty arbitrarily, if this RFC is accepted and we're happy with this number, IANA registration sounds like the right thing to pursue.

In terms of the UDS path I guess we'd need to first determine if we want plugins to be composable or if we want it to be a singleton. If we go the multiple-plugins route:
```
$os-specific-runtime-dir/wrpc/wasi/$name.sock
```
e.g.
```
/run/user/$uid/wrpc/wasi/http.sock
/run/user/$uid/wrpc/wasi/keyvalue.sock
/run/user/$uid/wrpc/myorg/mypackage.sock
```
seems like a reasonable default for wRPC-based plugins on UDS.